pkg/plugins/runtime/gateway/filter_chain_generator.go

```diff
@@ -104,31 +122,55 @@ func (g *HTTPSFilterChainGenerator) Generate(
 					return nil, nil, errors.Errorf("duplicate TLS certificate %q", secret.Name)
 				}
 
-				hostResources.Add(NewResource(secret.Name, secret))
+				resource := NewResource(secret.Name, secret)
+
+				hostResources.Add(resource)
+
+				downstream.CommonTlsContext.TlsCertificateSdsSecretConfigs = append(
+					downstream.CommonTlsContext.TlsCertificateSdsSecretConfigs, envoy_tls_v3.NewSecretConfigSource(resource.Name),
+				)
 			}
 
 		case mesh_proto.MeshGateway_TLS_PASSTHROUGH:
+			builder.Configure(
+				envoy_listeners.MatchServerNames(host.Hostname),
+			)
+
 			// TODO(jpeach) add support for PASSTHROUGH mode.
 			return nil, nil, errors.Errorf("unsupported TLS mode %q", host.TLS.GetMode())
 
-		default:
-			return nil, nil, errors.Errorf("unsupported TLS mode %q", host.TLS.GetMode())
-		}
+		case mesh_proto.MeshGateway_TLS_NONE:
+			if !info.Listener.CrossMesh {
+				return nil, nil, errors.Errorf("unsupported TLS mode %q", mode)
+			}
 
-		builder := newFilterChain(ctx, info)
+			// We don't match on the SNI here since it won't be the hostname
+			// here but rather an SNI based on the kuma.io/service
+			builder.Configure(envoy_listeners.MatchApplicationProtocols("kuma"))
 
-		builder.Configure(
-			envoy_listeners.MatchTransportProtocol("tls"),
-			envoy_listeners.MatchServerNames(host.Hostname),
-		)
+			otherMeshes := ctx.Mesh.Resources.OtherMeshes().Items
 
-		downstream := newDownstreamTypedConfig()
+			// We include all mesh certificates concatenated together
+			identity, ca, err := ctx.ControlPlane.Secrets.GetForGatewayListener(ctx.Mesh.Resource, info.Proxy.Dataplane, otherMeshes)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "couldn't generate mesh cert for gateway")
+			}
 
-		// If we generated any secrets, attach their references to the downstream validation context.
-		for _, s := range hostResources.ListOf(envoy_resource.SecretType) {
-			downstream.CommonTlsContext.TlsCertificateSdsSecretConfigs = append(
-				downstream.CommonTlsContext.TlsCertificateSdsSecretConfigs, envoy_tls_v3.NewSecretConfigSource(s.Name),
-			)
+			resources := xds_generator.CreateSecretResources(ctx.Mesh.Resource, identity, []xds_secrets.MeshCa{ca})
+
+			// We generate a downstream context using the envoy helpers
+			// and we don't want to match a SAN because it can be any mesh
+			downstream, err = envoy_tls_v3.CreateDownstreamTlsContext(ca.Mesh, ctx.Mesh.Resource)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "couldn't generate downstream tls context for gateway")
+			}
+
+			downstream.CommonTlsContext.GetCombinedValidationContext().DefaultValidationContext.MatchSubjectAltNames = nil
```

Comment:
technically you could put matchers for other meshes so
```
prefix: spiffee://mesh1
prefix: spiffee://mesh2
...
```
but probably not worth it, since it will be checked in TrafficPermission (RBAC filter)

Reply:
I'd to take a look at this in another PR.